[MRG] Support for Out-of-Bag cross validation: Add IdentitySplitter() and oob_score()

multimeric opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #23382


#### What does this implement/fix? Explain your changes.

* Add `IdentitySplitter()`, which is a dummy cross-validation splitter that just returns the training data. This can be used to implement cross validation for OOB
* Add `oob_score()`, a scorer function that delegates to the `RandomForest` oob score
* Added general tests for both, and for the integration of the two together in a CV context

#### Any other comments?
The intended usage of these new objects together is as follows:
```python
from sklearn.metrics import oob_score
from sklearn.model_selection import GridSearchCV, IdentitySplitter
from sklearn.ensemble import RandomForestClassifier

cv = GridSearchCV(
    RandomForestClassifier(oob_score=True, random_state=0),
    {"n_estimators": [1, 20, 100]},
    cv=IdentitySplitter(),
    scoring=oob_score,
)
results = cv.fit(X, y)
```

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

multimeric committed:
Add IdentitySplitter() and oob_score()

multimeric committed:
Add docstrings for new public entities

multimeric committed:
Add back in terminating newline

multimeric committed:
Update the whatsnew

multimeric committed:
Add credit in whatsnew

multimeric committed:
Flake 8

multimeric committed:
More flake 8

multimeric committed:
More flake 8

multimeric committed:
Run black

multimeric committed:
Whatsnew doc fixes

multimeric committed:
Fix test

multimeric committed:
Fix tests

multimeric committed:
Fix docstrings

multimeric committed:
Remove mistaken changelog entry

multimeric committed:
Pin random state for make_classification for reproducibility

multimeric commented:
@glemaitre, @ogrisel, @thomasjpfan (I'm following this list: https://github.com/scikit-learn/scikit-learn/wiki/Available-reviewers)

thomasjpfan commented on sklearn/metrics/_scorer.py:
```diff
@@ -771,6 +771,42 @@ def __getitem__(self, item):
+    OOB : float
+        Out-of-bag score.
+    """
+    if hasattr(estimator, "oob_score_"):
+        return estimator.oob_score_
```
If a `RandomForestClassifer` is in a `Pipeline`, then getting the attribute will be more complicated.

We can special case for `Pipeline`, but that will not work generically. There can be third party meta-estimators that can wrap a `RandomForestClassifer` that we can not special case for.

multimeric replied:
Is there no "get final estimator" method or function that is guaranteed to return the actual trained model from a pipeline that we can use? Ideally the special case won't have to live in the scorer but in a more re-usable place.

thomasjpfan replied:
One can special case for a `Pipeline` by doing `pipe[-1].oob_score_`. But that does not take into account all the different ways a third party estimator or meta-estimator may wrap a `RandomForest`.

Not being generic enough suggest that this function is better as an example than a function in the library. A user can already provide their own custom `oob_score` callable that works with their custom estimator.

multimeric replied:
I would strongly prefer this doesn't become an example, it will see much less use if it is.

If the meta estimators can't access estimator attributes, how do we access the `feature_importance_` when using something like RFECV? Surely there is a mechanism for this. Also I note that there is positive response to #23400 which might mean this is generalizable to several estimator classes.

multimeric replied:
Also, if you don't like it as a global class, how about as a mixin for the forest-based classifiers, such that you can
```python
cv = GridSearchCV(
    RandomForestClassifier(oob_score=True, random_state=0),
    {"n_estimators": [1, 20, 100]},
    cv=RandomForestClassifier.oob_cv,
    scoring=RandomForestClassifier.oob_scorer,
)
results = cv.fit(X, y)
```

thomasjpfan replied:
That could work for `cv`, but it is not standard practice for `sklearn`. The scorer would still have issues. The idea is to work with pipelines:

```python
clf = Pipeline(
    steps=[("preprocessor", preprocessor), ("classifier", RandomForestClassifier(oob_score=True))]
)

cv = GridSearchCV(
    clf,
    {"classifier__max_features": ["sqrt", "log2"]},
    cv=RandomForestClassifier.oob_cv,
    scoring=lambda est, X, y: est[-1].oob_score_
)
results = cv.fit(X, y)
```

Using `RandomForestClassifier.oob_scorer` does not work because it does not know it is in a Pipeline.

What if someone has a meta-estimator at the end, lets say a `VotingClassifier` and bags them?

```python
voting = VotingClassifier([
    BaggingClassifer(LogisticRegression(), oob_score=True),
    RandomForestClassifier(oob_score=True),
])
clf = Pipeline(
    steps=[("preprocessor", preprocessor), ("classifier", voting)]
)

def mean_oob_score_(est, X, y):
    # est is a pipeline with a voting classifer at the end.
    return np.mean([e.oob_score_ in est[-1].estimators_])

cv = GridSearchCV(
    clf,
    {"classifier__max_features": ["sqrt", "log2"]},
    cv=RandomForestClassifier.oob_cv,
    scoring=mean_oob_score_
)
results = cv.fit(X, y)
```

thomasjpfan replied:
> If the meta estimators can't access estimator attributes, how do we access the feature_importance_ when using something like RFECV?

[RFECV](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFECV.html#sklearn.feature_selection.RFECV) has a `importance_getter` allowing a user to define a custom way to get feature importances.

multimeric replied:
Right so could I not just define an optional `score_getter` that defaults to `oob_score_`? I get that there are a few less estimators that have a built-in score than have feature importance, but it applies to all the forest classifiers I think. Plus XGBoost externally.

thomasjpfan replied:
There are two user stories I am concerned about:

### Going from RandomForest -> RandomForest in Pipeline

Let's say this PR is merged so users can do this:

```python
cv = GridSearchCV(
    RandomForestClassifier(...), ...
    cv=IdentitySplitter(), scoring=oob_score
)
```

If they switch to a pipeline because of categorical data, they would need to define their own custom `oob_score`. We can work around this by special casing pipeline, but I am not happy with special casing if there is an alternative.

### Unintended usage

Imagine a user only setting `scoring=oob_score` and not the `cv=IdentitySplitter` or the other way around.

## Possible way forward

Even with this PR, we would need an example to showcase the feature. Moving forward, I think we can have an example that defines a custom splitter and function and use them to show case oob_score for validation. As a follow up, we can decide on moving the Splitter and function into the library.

multimeric replied:
Okay, fair enough. If you would like to see this as an example then I'll submit another PR which refactors this into one.

thomasjpfan replied:
Also, I suggest having this PR become the example PR. Maintainer time and reviewer time is limited, having one PRs helps focus attention.

multimeric replied:
Understood. But I don't want to lose the work I've done, and I don't think I can switch branches etc.

thomasjpfan replied:
Since including the feature has not been decided yet, I think we can add an example on top of this PR with `IdentitySplitter` and `oob_score` being a part of the library.

If we decide to only have an example, then we can refactor this PR into a new one that only has the example.

multimeric replied:
Resolving since none of this matters now that the classes aren't in the PR.

thomasjpfan reviewed (commented):
Thank you for the PR.

Before going into a more technical review, we discussion the feature's inclusion in the original issue: https://github.com/scikit-learn/scikit-learn/issues/23382#issuecomment-1146817142

multimeric committed:
Add oob cross validation example

multimeric committed:
Fix example

multimeric committed:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into oob-cv

multimeric commented:
Okay, I've now added an example of this feature. It **doesn't** rely on the new classes I added, in case it is decided that these classes are not wanted in the repo. If this is decided I will remove them from the PR and keep only the example.

multimeric committed:
Black

multimeric commented:
Here's the rendered version of the example: https://output.circle-artifacts.com/output/job/3166928d-f107-471f-80c5-ea49d5d16d98/artifacts/0/doc/auto_examples/ensemble/plot_oob_cross_validation.html

thomasjpfan commented on examples/ensemble/plot_oob_cross_validation.py:
```diff
@@ -0,0 +1,87 @@
+:math:`k` passes in :math:`k`-fold cross validation [1]_.
+
+In this example we perform greedy sequential feature selection with a small number of
+features, and compare the runtime for out-of-bag error and cross validation error.
```
This example runs for 33 seconds, that is usually a bit long for an example. I do not think we need to include the `cv=5`.

multimeric committed:
Convert example to parameter selection, add additional context, remove
class based edits

multimeric committed:
Update whatsnew

multimeric commented:
Rendered example: https://output.circle-artifacts.com/output/job/5a965e7b-1ff0-487a-a346-a48e07642613/artifacts/0/doc/auto_examples/ensemble/plot_oob_cross_validation.html

glemaitre commented:
As mentioned here: https://github.com/scikit-learn/scikit-learn/issues/23382#issuecomment-1152165906, I would be more in favor of repurposing the already existing example and making it narrative instead of adding a new example. One of the reasons is that we have already too many examples and we should avoid adding a new one when we can improve an existing one (and old one).

multimeric commented:
Considering the reduced scope of what you would accept, I don't have any particular interest in finishing this PR, but I will keep the issue #23382 open because I believe it's still an resolved need in sklearn.
